cdap-app-fabric/src/main/java/io/cdap/cdap/app/job/SupportBundleJob.java

```diff
@@ -0,0 +1,471 @@
+                          runRecordDetail != null && runRecordDetail.getStopTs() != null
+                              ? runRecordDetail.getStopTs()
+                              : DateTime.now().getMillis());
+                  retryServiceMap.remove(runId);
+                } catch (Exception e) {
+                  LOG.warn(
+                      String.format("Retried three times for this metrics with run id %s ", runId),
+                      e);
+                  queueTaskAfterFailed(runId, subMetricCalculateTask, basePath);
+                }
+                return new Pair<>(runRecordDetail, metrics);
+              });
+      if (!retryServiceMap.containsKey(runId)) {
+        futureMetricsList.add(futureMetricsService);
+      }
+    }
+    completeProcessing(futurePipelineInfoList, pipelineTaskMap, basePath, "pipelineinfo");
+    completeProcessingForMetrics(futureMetricsList, subMetricCalculateTask, basePath);
+  }
+
+  /** Generates run log file */
+  public void runLogProcessing(String basePath, String appPath) {
+    List<Future> futureRunLogList = new ArrayList<>();
+    Map<String, JsonObject> taskMap = new HashMap<>();
+    Queue<RunRecordDetail> runRecordDetailQueue = new LinkedList<>();
+    for (RunRecordDetail runRecordDetail : runRecordDetailList) {
+      runRecordDetailQueue.offer(runRecordDetail);
+    }
+    while (!runRecordDetailQueue.isEmpty()) {
+      RunRecordDetail runRecordDetail = runRecordDetailQueue.poll();
+      String runId = runRecordDetail.getProgramRunId().getRun();
+      JsonObject task = initializeTask(runId + "-log", "PipelineRunLogSupportBundleTask");
+      taskMap.put(runId, task);
+      Future<String> futureRunLogService =
+          executor.submit(
+              () -> {
+                try {
+                  updateTask(task, basePath, "IN_PROGRESS");
+                  LoggingContext loggingContext =
+                      LoggingContextHelper.getLoggingContextWithRunId(
+                          runRecordDetail.getProgramRunId(), null);
+                  supportBundlePipelineRunLogTask.generateRunLogFile(
+                      cConf, logReader, loggingContext, appPath, runId + "-log.txt");
+                  retryServiceMap.remove(runId);
+                } catch (Exception e) {
+                  LOG.warn(
+                      String.format(
+                          "Retried three times for this run id %s to generate run log ", runId),
+                      e);
+                  queueTaskAfterFailed(runId, task, basePath);
+                }
+                return runId;
+              });
+      if (!retryServiceMap.containsKey(runId)) {
+        futureRunLogList.add(futureRunLogService);
+      }
+    }
+    completeProcessing(futureRunLogList, taskMap, basePath, "runlog");
+  }
+
+  /** Generates run info file */
+  public void runInfoProcessing(
+      ApplicationRecord applicationRecord, String basePath, String appPath) {
```

Comment:
Again, any task-specific code must be in task, job should do coordination and provide utility functions